File: sqrt.py
class Solution:
    # @param x, an integer
    # @return an integer
    def sqrt(self, x):
        start = 0
        end = x
        while start <= end:
        	mid = start + end
        	mid /= 2
        	if mid * mid <= x and (mid + 1) * (mid + 1) > x :
        		return int(mid)
        	elif mid * mid <= x:
        		start = mid + 1
        	else:
        		end = mid - 1

#use division to avoid multiplication overflow

s = Solution()
print s.sqrt(10) == 3
print s.sqrt(0.9) == 0
print s.sqrt(100) == 10

File: JAVA/sqrt.java
public class Solution {
    public int sqrt(int x) {
        if (x < 1)
            return 0;
        if (x < 4)
            return 1;
            
        float quotient;
        int low = 0, high = x, mid;
        while (low < high){
            mid = (low + high) / 2;
            if (x / mid >= mid && x / (mid + 1) < (mid + 1))
                return mid;
            else if (x / mid >= mid)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return low;
    }
}